File: ec_production/ec_production/report/voucher_report/voucher_report.py
# Copyright (c) 2026, SkylineBiz Private Limited and contributors
# For license information, please see license.txt

import frappe


def execute(filters=None):
    columns = [
        {"label": "Date", "fieldname": "date", "fieldtype": "Date", "width": 120},
        {
            "label": "Employee",
            "fieldname": "employee",
            "fieldtype": "Data",
            "width": 250,
        },
        {
            "label": "EC Lot",
            "fieldname": "ec_lot",
            "fieldtype": "Link",
            "options": "EC Lot",
            "width": 150,
        },
        {
            "label": "Item",
            "fieldname": "item",
            "fieldtype": "Link",
            "options": "Item",
            "width": 180,
        },
        {
            "label": "Operation",
            "fieldname": "operation",
            "fieldtype": "Link",
            "options": "Operation",
            "width": 150,
        },
        {"label": "Qty", "fieldname": "qty", "fieldtype": "Float", "width": 100},
        {"label": "Rate", "fieldname": "rate", "fieldtype": "Currency", "width": 120},
        {
            "label": "Amount",
            "fieldname": "amount",
            "fieldtype": "Currency",
            "width": 120,
        },
    ]

    conditions = []
    values = {}

    if filters.get("from_date"):
        conditions.append("pl.date >= %(from_date)s")
        values["from_date"] = filters.get("from_date")

    if filters.get("to_date"):
        conditions.append("pl.date <= %(to_date)s")
        values["to_date"] = filters.get("to_date")

    employees = filters.get("employee") or []

    if employees:
        conditions.append("emp.name IN %(employees)s")
        values["employees"] = tuple(employees)

    where_clause = ""
    if conditions:
        where_clause = "WHERE " + " AND ".join(conditions)

    data = frappe.db.sql(
        f"""
        SELECT
            pl.date AS date,
            CONCAT(pli.employee, ': ', emp.employee_name) AS employee,
            pli.ec_lot AS ec_lot,
            pli.item AS item,
            pli.operation AS operation,
            pli.qty AS qty,
            pli.rate AS rate,
            pli.amount AS amount
        FROM `tabEC Process Lot Item` pli
        INNER JOIN `tabEC Process Lot` pl
            ON pl.name = pli.parent
        LEFT JOIN `tabEmployee` emp
            ON emp.name = pli.employee
        {where_clause}
        ORDER BY
            pl.date DESC,
            emp.employee_name
        """,
        values,
        as_dict=True,
    )

    return columns, data

File: ec_production/ec_production/report/voucher_report/voucher_report.js
frappe.query_reports["Voucher Report"] = {
    filters: [
        {
            fieldname: "from_date",
            label: __("From"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.month_start()
        },
        {
            fieldname: "to_date",
            label: __("To"),
            fieldtype: "Date",
            reqd: 1,
            default: frappe.datetime.month_end()
        },
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "MultiSelectList",
            get_data: function(txt) {
                return frappe.db.get_link_options("Employee", txt);
            }
        }
    ],

    onload: function(report) {
        if (frappe.route_options) {

            if (frappe.route_options.employee) {
                report.set_filter_value(
                    "employee",
                    Array.isArray(frappe.route_options.employee)
                        ? frappe.route_options.employee
                        : [frappe.route_options.employee]
                );
            }

            report.set_filter_value("from_date", frappe.route_options.from_date);
            report.set_filter_value("to_date", frappe.route_options.to_date);

            frappe.route_options = null;
            report.refresh();
        }
    }
};